txnsync/peer.go

```diff
@@ -228,26 +228,28 @@ func (p *Peer) selectPendingTransactions(pendingTransactions []transactions.Sign
 			continue
 		}
 
-		p.lastTransactionSelectionGroupCounter = pendingTransactions[grpIdx].GroupCounter
-
 		if windowSizedReached {
 			hasMorePendingTransactions = true
 			break
 		}
 		selectedTxns = append(selectedTxns, pendingTransactions[grpIdx])
 		selectedTxnIDs = append(selectedTxnIDs, txID)
 
-		// calculate the total size of the transaction group.
-		for txidx := range pendingTransactions[grpIdx].Transactions {
-			encodingBuf := protocol.GetEncodingBuf()
-			accumulatedSize += len(pendingTransactions[grpIdx].Transactions[txidx].MarshalMsg(encodingBuf))
-			protocol.PutEncodingBuf(encodingBuf)
-		}
+		// add the size of the transaction group
+		accumulatedSize += pendingTransactions[grpIdx].EncodedLength
+
 		if accumulatedSize > windowLengthBytes {
 			windowSizedReached = true
 		}
 	}
 
+	if grpIdx == len(pendingTransactions) {
+		grpIdx--
+	}
+	if grpIdx >= 0 {
+		p.lastTransactionSelectionGroupCounter = pendingTransactions[grpIdx].GroupCounter
```

Comment:
I think it would be clearer to handle the empty case directly, I was a bit confused by the `grpIdx--` and `grpIdx >= 0` logic.

Something like `grpIdx != 0 && grpIdx == len(pendingTransactions)`, then access `grpIdx - 1`.

Alternatively update `p.lastTransactionSelectionGroupCounter` on each iteration of the loop.

Reply:
I wanted to avoid `p.lastTransactionSelectionGroupCounter` on each loop iteration. I'll update the predicate as you've suggested.

Reply:
I think that this expression wasn't good enough. Let me run few more test on this one and I'll ask for another review once I'm done.

Reply:
This feels like premature optimization, assigning `p.lastTransactionSelectionGroupCounter` is simpler.

Reply:
I agree that just assigning the `p.lastTransactionSelectionGroupCounter` is truly simpler. However, I found that this method was taking a considerable amount of the total txnsync time, so I'm trying to optimize it. Moving operations out of the loop is my first go-to solution here.